### DISCUSSION TITLE
Does react-table support readonly data?

### DISCUSSION THREAD

**Discussion by Pavinati** [TanStack/table#3648] [Q&A]
Seems like the table data prop is forced to be mutable. is it required to be mutable or is there a way around this?

```typescript
type SomeRecord = Readonly<....>;

const columns: Column<SomeRecord>[] = [...];

type Props {
  className?: string;
  data: readonly SomeRecord[];
}

const MyTable = ({ className, data }: Props) => {
  return <Table className={className} columns={columns} data={data} />;
};
```
results in
```
The type 'readonly Readonly<SomeRecord>[]' is 'readonly' and cannot be assigned to the mutable type 'Readonly<SomeRecord>[]'.  TS4104
```

**Comment by davidebriani** [comment]
Up. Supporting readonly TypeScript objects would be nice. My use-case is that I get GraphQL data from Relay, which provides auto-generated TypeScript types with `readonly` signatures.

**Comment by pcorpet** [comment]
I've sent a PR https://github.com/TanStack/table/pull/4909

**Comment by deammer** [comment]
This would be particularly helpful when using `gql-tada`, which returns `readonly` types:

```ts
import { useReactTable } from '@tanstack/react-table';
import { FragmentOf, graphql, readFragment } from '@/graphql';

const myFragment = graphql(`
  fragment MyFragment on Order {
    id
  }
`);

const MyReactComponent = (props: { orders: FragmentOf<typeof myFragment>[]) => {
  const items = readFragment(myFragment, props.orders);
  const table = useReactTable({
    data: items,
//  ^ the type (...) is 'readonly' and cannot be assigned to the mutable type (...)
  });
}

```

**Comment by rsslldnphy** [comment]
Just adding to the chorus - my ORM also returns read only types, which works everywhere except when using react-table, when I am forced to either typecast or copy the data. Making the argument a readonly type would mean that both mutable and readonly data can be used; in its current state it supports mutable only.
